src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -38,31 +46,110 @@ const CompressionCodecPtr & GinIndexCompressionFactory::zstdCodec()
 {
-    return rowids.contains(row_id);
+    std::vector<UInt32> deltas = encodeDeltaScalar(rowids);
+
+    /// FastPFOR requires the output buffer to be "big enough", so +20% buffer is our attempt to comply with that.
+    std::vector<UInt32> compressed(static_cast<size_t>(std::ceil(deltas.size() * 1.2)));
+    size_t compressed_size = compressed.size();
+    if (deltas.size() < FASTPFOR_THRESHOLD)
+    {
+        std::memcpy(compressed.data(), deltas.data(), sizeof(UInt32) * deltas.size());
+        compressed_size = deltas.size();
+    }
+    else
+    {
+        codec()->encodeArray(deltas.data(), deltas.size(), compressed.data(), compressed_size);
+    }
+
+    UInt64 written_bytes = 0;
+
+    UInt64 num_deltas = deltas.size();
+    writeVarUInt(num_deltas, buffer);
+    written_bytes += getLengthOfVarUInt(num_deltas);
+
+    writeVarUInt(compressed_size, buffer);
+    written_bytes += getLengthOfVarUInt(compressed_size);
+
+    buffer.write(reinterpret_cast<char *>(compressed.data()), compressed_size * sizeof(UInt32));
+    written_bytes += compressed_size * sizeof(UInt32);
+
+    return written_bytes;
 }
 
-void GinIndexPostingsBuilder::add(UInt32 row_id)
+GinIndexPostingsListPtr GinIndexPostingListDeltaPforSerialization::deserialize(ReadBuffer & buffer)
 {
-    rowids.add(row_id);
+    size_t num_deltas = 0;
+    size_t compressed_size = 0;
+    readVarUInt(num_deltas, buffer);
+    readVarUInt(compressed_size, buffer);
+
+    std::vector<UInt32> compressed(compressed_size);
+    buffer.readStrict(reinterpret_cast<char *>(compressed.data()), compressed_size * sizeof(UInt32));
+
+    std::vector<UInt32> deltas(num_deltas);
+    if (deltas.size() < FASTPFOR_THRESHOLD)
+    {
+        std::memcpy(deltas.data(), compressed.data(), sizeof(UInt32) * deltas.size());
+    }
+    else
+    {
+        codec()->decodeArray(compressed.data(), compressed_size, deltas.data(), num_deltas);
+    }
+
+    decodeDeltaScalar(deltas);
+
+    GinIndexPostingsListPtr postings_list = std::make_shared<GinIndexPostingsList>();
+    postings_list->addMany(deltas.size(), deltas.data());
+    return postings_list;
 }
 
-UInt64 GinIndexPostingsBuilder::serialize(WriteBuffer & buffer)
+std::shared_ptr<FastPForLib::IntegerCODEC> GinIndexPostingListDeltaPforSerialization::codec()
 {
-    rowids.runOptimize();
+    static thread_local FastPForLib::CODECFactory factory;
+    static std::shared_ptr<FastPForLib::IntegerCODEC> codec = factory.getFromName(FASTPFOR_CODEC_NAME);
+    return codec;
+}
 
-    const UInt64 cardinality = rowids.cardinality();
+std::vector<UInt32> GinIndexPostingListDeltaPforSerialization::encodeDeltaScalar(const roaring::Roaring & rowids)
+{
+    const UInt64 num_rowids = rowids.cardinality();
+    std::vector<UInt32> deltas(num_rowids);
+    UInt32 prev = 0;
+    for (size_t i = 0; const UInt32 rowid : rowids)
+    {
+        prev = rowid;
+        deltas[i] = rowid - prev;
```

Comment:
in this case deltas are always zero since it becomes `rowid - rowid`. I change the order again, it could be `curr(rowid) - prev`

Reply:
fast test does not because `ENABLE_FASTPFOR` is not enabled in CI build